README.md
# ScreenshotCrew

[![Node.js CI](https://github.com/jadiagaurang/ScreenshotCrew/actions/workflows/node.js.yml/badge.svg)](https://github.com/jadiagaurang/ScreenshotCrew/actions/workflows/node.js.yml)
[![Docker Image CI](https://github.com/jadiagaurang/ScreenshotCrew/actions/workflows/docker-image.yml/badge.svg)](https://github.com/jadiagaurang/ScreenshotCrew/actions/workflows/docker-image.yml)
[![Docker pulls](https://img.shields.io/docker/pulls/jadiagaurang/screenshotcrew.svg?logo=docker)](https://hub.docker.com/r/jadiagaurang/screenshotcrew/)

ScreenshotCrew is an open-source screenshot as a service using [Puppeteer headless browser](https://github.com/puppeteer/puppeteer). Using Node.js, [Puppeteer](https://pptr.dev/) and other plugins, ScreenshotCrew provides utility tool to create screenshot from any web page.

## Motivation

* https://screenshotlayer.com/
* https://urlbox.io/
* https://browshot.com/
* https://screenshotmachine.com/

## Demo

[https://screenshotcrew.com/](https://screenshotcrew.com/)

## Installation

```base
npm install
npm update
```

## Code Example

### PROD

```base
npm start
```

### LOCAL DEV

```base
npm run-script dev
```

## API Reference

```bash
curl --location --request GET "https://screenshotcrew.com/api/capture?url=https://gaurangjadia.com"
```

```bash
curl --location --request POST "https://screenshotcrew.com/api/capture" --header 'Content-Type: application/json' --data-raw '{
    "url": "https://gaurangjadia.com"
}'
```

## Tests

```base
npm test
```

## License

Please see the [license file](https://github.com/jadiagaurang/ScreenshotCrew/blob/main/LICENSE) for more information.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2021-05-05
### Added
- Initial Commit with MVP Support

### Changed

### Removed

[1.0.0]: https://github.com/jadiagaurang/ScreenshotCrew/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/jadiagaurang/ScreenshotCrew/releases/tag/v1.0.0

tests/index.spec.js
"use strict";

const supertest = require("supertest");
const assert = require("assert");
const app = require("../app");

describe("GET /", function() {
    it("Index StatusCode 200 OK", function(done) {
        supertest(app)
		.get("/")
		.expect(200)
		.end(function(err, res){
			if (err) done(err);
			done();
		});
	});
});

tests/shotbot.spec.js
"use strict";

const supertest = require("supertest");
const assert = require("assert");
const expect = require("chai").expect;
const app = require("../app");

describe("GET /api/capture", function() {
    it("Capture StatusCode 200 OK with screenshotPath not null", function(done) {
        supertest(app)
		.get("/api/capture")
		.query({ "url": "https://gaurangjadia.com" })
		.set("accept", "application/json")
		.expect(200)
		.expect("Content-Type", /json/)
		.end(function(err, res){
			if (err) {
				return done(err);
			}

			expect(res.body.screenshotPath.length).greaterThan(0);

			return done();
		});
	});
});

describe("POST /api/capture", function() {
    it("Capture StatusCode 200 OK with screenshotPath not null", function(done) {
        supertest(app)
		.post("/api/capture")
		.send({ "url": "https://gaurangjadia.com" })
		.set("accept", "application/json")
		.expect(200)
		.expect("Content-Type", /json/)
		.end(function(err, res){
			if (err) {
				return done(err);
			}

			expect(res.body.screenshotPath.length).greaterThan(0);

			return done();
		});
	});
});

app.js
"use strict";

// Load external plugins
const dotenv = require("dotenv");
const express = require("express");
const fs = require("fs");
const path = require("path");
const favicon = require("serve-favicon");
const logger = require("morgan");
const cookieParser = require("cookie-parser");
const bodyParser = require("body-parser");

// Load local plugins
const winston = require("./src/logger").winston;
const routes = require("./routes/index");
const shotbot = require('./routes/shotbot');

// ExpressJS App Setup
var app = express();

// Load Environment Variables from the Config File
const args = process.argv;
const fileName = path.dirname(fs.realpathSync(__filename));

// Work on the Args
if (args.length > 2) {
    var objEnv = args[2];
    
    if (objEnv === "prod") {
        dotenv.config({ path: path.join(fileName, "./config/prod.env") });
    }
    else {
        dotenv.config({ path: path.join(fileName, "./config/dev.env") });
    }
}
else {
    dotenv.config({ path: path.join(fileName, "./config/dev.env") });
}

// View Engine Setup
app.set("views", path.join(__dirname, "views"));
app.set("view engine", "ejs");

// Middlewave Setup
app.use(favicon(__dirname + "/public/images/favicon.ico"));
app.use(logger("dev"));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, "public")));

// WebApp and API Route Setup
app.use("/", routes);
app.use("/api", shotbot);

// Catch 404 and forward to error handler
app.use(function (req, res, next) {
    var err = new Error("Not Found");
    err.status = 404;
    next(err);
});

// error handlers
// development error handler will print stacktrace
if (app.get("env") === "development") {
    app.use(function (err, req, res, next) {
        res.status(err.status || 500);
        res.render('error', {
            title: "Error",
            message: err.message,
            error: err
        });
    });
}

// production error handler no stacktraces leaked to user
app.use(function (err, req, res, next) {
    res.status(err.status || 500);
    res.render('error', {
        title: "Error",
        message: err.message,
        error: {}
    });
});

// Port Setup fallback to 3000
app.set("port", process.env.PORT || 3000);

// ExpressJS Server Setup
var server = app.listen(app.get("port"), function () {
    var meLogger = winston(process.env.LOG_LEVEL);

    meLogger.info('Express server listening on port ' + server.address().port);
});

// Return ExpressJS Server
module.exports = server

src/utility.js
"use strict";

const Memcached = require("memcached");

module.exports = class Utility {
	static isBlank (obj) {
		return (!obj || obj === "");
	}

	static async getCachedResult (strKey) {
		return new Promise((resolve, reject) => {
			var memcacheHost = process.env.SERVER_MEMCACHE_HOST;
			var memcachePort = process.env.SERVER_MEMCACHE_PORT;

			if (!Utility.isBlank(memcacheHost) && !Utility.isBlank(memcachePort)) {
				//Make CacheKey
				var varCacheKey = "Host=" + process.env.SERVER_HOST + "&Method=getCachedResult&strKey=" + strKey;

				//Set up Memcache Client
				var dcOptions = {
					timeout: 5000   //the time after which Memcached sends a connection timeout (in milliseconds).
				};
				var memcachedClient = new Memcached(memcacheHost + ":" + memcachePort, dcOptions);

				memcachedClient.gets(varCacheKey, function (err, data) {
					// If data received from memcached and isCache flag active.
					if (!Utility.isBlank(data) && !Utility.isBlank(data[varCacheKey])) {
						resolve(data[varCacheKey]);
					}
					else {
						resolve(null);
					}
				});
			}
			else {
				resolve(null);
			}
		});
	}

	static async setCachedResult (strKey, data) {
		return new Promise((resolve, reject) => {
			var memcacheHost = process.env.SERVER_MEMCACHE_HOST;
			var memcachePort = process.env.SERVER_MEMCACHE_PORT;

			if (!Utility.isBlank(memcacheHost) && !Utility.isBlank(memcachePort)) {
				//Make CacheKey
				var varCacheKey = "Host=" + process.env.SERVER_HOST + "&Method=setCachedResult&strKey=" + strKey;

				//Set up Memcache Client
				var dcOptions = {
					timeout: 5000   //the time after which Memcached sends a connection timeout (in milliseconds).
				};
				var memcachedClient = new Memcached(memcacheHost + ":" + memcachePort, dcOptions);

				memcachedClient.set(varCacheKey, data, (60 * 1000), function (err) {
					if (!Utility.isBlank(err)) {
						resolve(false);
					}
					else {
						resolve(true);
					}
				});
			}
			else {
				resolve(false);
			}
		});
	}

	static delay(timeout) {
		return new Promise((resolve) => {
			setTimeout(resolve, timeout);
		});
	}
}

src/shotbot.js
"use strict";

const puppeteer = require("puppeteer");
const fs = require("fs");
const path = require("path");
const { v4: uuidv4 } = require("uuid");
const { URL } = require("url");
const AWS = require("aws-sdk");

const winston = require("./logger").winston;
const util = require("./utility.js");

module.exports = class ShotBot {
	//Default Constructor
	constructor(strURL) {
		var me = this;
		
		me.logger = winston(process.env.LOG_LEVEL);

		if (!util.isBlank(strURL)) {
			this.objURL = new URL(strURL);
		}
		else {
			this.objURL = null;
		}
		
		this.dimensions = {
			width: 0,
			height: 0
		};

		//Default Values
		this.requestId = uuidv4();
		this.screenshotPath = path.join(path.dirname(fs.realpathSync(__filename)), "../screenshots/");
		this.appServerURL = process.env.APP_SERVER;
	}

	async getScreenshot() {
		var me = this;

		me.logger.info("getScreenshot : " + this.objURL);

		const args = ["--disable-setuid-sandbox", "--no-sandbox", "--disable-extensions"];
		const options = {
			args,
			headless: true
		};
		const browser = await puppeteer.launch(options);

		try {
			//Puppeteer Page
			const page = await browser.newPage();

			//Setup Page
			await this.setup(page);

			let imageName = await this.takeScrenshot(page);
			let imageCloudFrontURL = await this.uploadScreenshot(imageName);

			//Close Browser Object
			browser.close();

			if (!util.isBlank(imageCloudFrontURL)) {
				return {
					"screenshotPath": imageCloudFrontURL["FilePath"]
				};
			}
			else {
				return {
					"screenshotPath": this.appServerURL + "/api/screenshot/" + imageName
				};
			}
		}
		catch(ex) {
			browser.close();

			throw ex;
		}
	}

	//Intitial operation required for the segmentation to run.
	async setup(page) {
		var me = this;

		me.logger.info("Setup : " + this.objURL);

		try {
			//Set Viewport
			page.setViewport({
				width: 1920,
				height: 1080,
				deviceScaleFactor: 1
			});

			//Setup UserAgent
			const userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.54 Safari/537.36";
			await page.setUserAgent(userAgent);

			await page.goto(this.objURL, {
				waitUntil: "networkidle2",
				timeout: 90000	//1.5 minutes
			});

			//Delay 1s
			await util.delay(1000);

			//Scroll to Bottom of the Page to unvail all the lazy/animated content of the page
			await this.scrollToBottom(page);
			await this.scrollToTop(page);

			//Delay 1s
			await util.delay(1000);

			// Get Page's Width and Height to debug wrong section data
			this.dimensions = await page.evaluate(() => {
				return {
					width: Math.max(
						1280,
						Math.max(document.body.scrollWidth, document.documentElement.scrollWidth),
						Math.max(document.body.offsetWidth, document.documentElement.offsetWidth),
						Math.max(document.body.clientWidth, document.documentElement.clientWidth)
					),
					height: Math.max(
						Math.max(document.body.scrollHeight, document.documentElement.scrollHeight),
						Math.max(document.body.offsetHeight, document.documentElement.offsetHeight),
						Math.max(document.body.clientHeight, document.documentElement.clientHeight)
					)
				};
			});
		}
		catch(ex) {
			me.logger.error(ex);
		}
	}

	async scrollToTop(page) {
		var me = this;

		me.logger.info("scrollToTop : " + this.objURL);

		await page.evaluate(function() {
			window.setTimeout(function () {
				window.scrollTo(0, 0);	//Back to x = 0, y = 0
			}, 250);
		});

		await util.delay(500);
	}

	async scrollToBottom(page) {
		var me = this;

		me.logger.info("scrollToBottom : " + this.objURL);

		// Scrolling to bottom to load all the lazy loading contents
		await page.evaluate(function() {
			//Figure out Max Height
			var maxHeight = Math.max(
				Math.max(document.body.scrollHeight, document.documentElement.scrollHeight),
				Math.max(document.body.offsetHeight, document.documentElement.offsetHeight),
				Math.max(document.body.clientHeight, document.documentElement.clientHeight)
			);

			function scrollToBottom(x, iInc, callBack) {
				if (maxHeight > x) {
					window.setTimeout(function () {
						window.scrollTo(0, x + iInc);
						scrollToBottom(x + iInc, iInc, callBack);
					}, 250);
				}
				else {
					if (callBack) {
						callBack();
					}
				}
			}

			scrollToBottom(0, Math.ceil(maxHeight / 10), function () {
				//TODO: Nothing
			});
		});

		//10 Iteration Only each of them will be 250 ms so, 2500 total
		await util.delay(2500);
	}

	async takeScrenshot(page) {
		var me = this;

		me.logger.info("takeScrenshot : " + this.objURL);

		//Create Screenshot Directory if doesn't exist?
		if (!fs.existsSync(this.screenshotPath)){
			fs.mkdirSync(this.screenshotPath);
			me.logger.info("takeScrenshot - Making Directory at " + this.screenshotPath);
		}

		let imageName = this.requestId + ".webp";
		let screenshotPath = this.screenshotPath + imageName;

		//FullPage is causing webpage to scale so, adding clip screenshot correctly!
		await page.screenshot({
			path: screenshotPath,
			type: "webp",	// jpeg, png or webp
			quality: 100,
			fullPage: true,
			omitBackground: false,
			encoding: "binary",	//	base64 or binary
			captureBeyondViewport: true,
		});

		return imageName;
	}

	async uploadScreenshot(imageName) {
		var me = this;

		if (!util.isBlank(process.env.AWS_ACCESS_KEY_ID) && 
			!util.isBlank(process.env.AWS_SECRET_ACCESS_KEY)) {

			me.logger.info("uploadScreenshot : " + this.objURL);

			let screenshotPath = this.screenshotPath + imageName;

			const s3 = new AWS.S3({
				accessKeyId: process.env.AWS_ACCESS_KEY_ID,
				secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
			});
		
			// Read content from the file
			let fileContent = fs.readFileSync(screenshotPath);
			let screenshotName = uuidv4();

			// Setting up S3 upload parameters
			const params = {
				Bucket: "",	// TODO: 
				Key: "screenshotcrew/" + screenshotName + ".webp",	// File name you want to save as in S3
				Body: fileContent
			};

			return new Promise(function (resolve, reject) {
				// Uploading files to the bucket
				s3.upload(params, function(err, data) {
					if (err) {
						reject(err);
					}

					resolve(data);
				});
			});
		}
		else {
			return Promise.resolve(null);
		}
	}

	fetchScreenshot(req, res, imageName) {
		var me = this;

		me.logger.info("fetchScreenshot: imageName = " + imageName);

		let imagePath = this.screenshotPath + imageName;
	
		//Check if the image exists
		fs.stat(imagePath, function(errOutter, stat) {
			if(errOutter) {
				res.writeHead(500, {"Content-Type": "application/json"});
				res.write(JSON.stringify({
					"Status": "ERROR",
					"Message": "File Not Found",
					"Detail": errOutter
				}));
				res.end();
			}
			else {
				//me.logger.info(stat);

				fs.readFile(imagePath, function(errInner, data) {
					if(errInner) {
						res.writeHead(500, {"Content-Type": "application/json"});
						res.write(JSON.stringify({
							"Status": "ERROR",
							"Message": "File Not Found",
							"Detail": errInner
						}));
						res.end();
					}
					else {
						res.writeHead(200, { "Content-Type": "image/webp" });
						res.end(data);
					}
				});
			}
		});
	}
}